doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,1773 @@
+only ever be defined as a class variable equal to `True` or `False`. If the
+assumption depends on the `.args` of the function somehow, define the <code
+class="docutils literal notranslate"><span
+class="pre">\_eval\_*assumption*</span></code> method.
+
+```
+
+In this example, it is not necessary to define `_eval_is_real()` because it is
+deduced automatically from the other assumptions, since `nonnegative -> real`.
+In general, you should avoid defining assumptions that the assumptions system
+can deduce automatically given its [known
+facts](assumptions-guide-predicates).
+
+```py
+>>> versin(1).is_real
+True
+```
+
+The assumptions system is often able to deduce more than you might think.
+For example, from the above, it can deduce that `versin(2*n*pi)` is zero when
+`n` is an integer.
+
+```py
+>>> n = symbols('n', integer=True)
+>>> versin(2*n*pi).is_zero
+True
+```
+
+It's always worth checking if the assumptions system can deduce something
+automatically before manually coding it.
+
+Finally, a word of warning: be very careful about correctness when coding
+assumptions. Make sure to use the exact
+[definitions](assumptions-guide-predicates) of the various assumptions, and
+always check that you're handling `None` cases correctly with the fuzzy
+three-valued logic functions. Incorrect or inconsistent assumptions can lead
+to subtle bugs. It's recommended to use unit tests to check all the various
+cases whenever your function has a nontrivial assumption handler. All
+functions defined in SymPy itself are required to be extensively tested.
+
+(custom-functions-evalf)=
+### Numerical Evaluation with `evalf()`
+
+<!-- TODO: this goes over the basics, but it might be useful to have a
+     separate guide dedicated to numerical evaluation. -->
+
+Here we show how to define how a function should numerically evaluate to a
+floating point {class}`~.Float` value, for instance, via `evalf()`.
+Implementing numerical evaluation enables several behaviors in SymPy. For
+example, once `evalf()` is defined, you can plot your function, and things
+like inequalities can evaluate to explicit values.
+
+If your function has the same name as a function in
+[mpmath](https://mpmath.org/doc/current/), which is the case for most
+functions included with SymPy, numerical evaluation will happen automatically
+and you do not need to do anything.
+
+If this is not the case, numerical evaluation can be specified by defining the
+method `_eval_evalf(self, prec)`, where `prec` binary precision of the input.
+The method should return the expression evaluated to the given precision, or
+`None` if this is not possible.
+
+```{note}
+
+The `prec` argument to `_eval_evalf()` is the *binary* precision, that is, the
+number of bits in the floating-point representation. This differs from the
+first argument to the `evalf()` method, which is the *decimal* precision, or
+`dps`. For example, the default binary precision of `Float` is 53,
```

Comment:
Perhaps worth mentioning the conversion functions `prec_to_dps` et al

Reply:
I'm not a fan of that function, because it loses precision (so to speak). Multiple binary prec's correspond to the same dps. Ideally evalf would just take `prec` as an optional argument as an alternative to dps (I can't remember if this has been discussed before).